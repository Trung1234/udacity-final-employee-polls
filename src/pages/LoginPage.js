import { connect } from "react-redux";
import { Navigate } from "react-router-dom";
import { useState } from "react";
import { handleLogin } from "../store/actions/authedUser";

const Login = ({ dispatch, loggedIn }) => {
  const [username, setUsername] = useState("tylermcginnis");
  const [password, setPassword] = useState("abc321");

  if (loggedIn) {
    // const urlParams = new URLSearchParams(window.location.search);
    // const redirectUrl = urlParams.get('redirectTo');
    return <Navigate to={"/"} />;
  }
  const handleUsername = (e) => {
    const value = e.target.value;
    setUsername(value);
  };

  const handlePassword = (e) => {
    const value = e.target.value;
    setPassword(value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(handleLogin(username, password));
    setUsername("");
    setPassword("");
  };

  return (
    <div>
      <h1 className="text-3xl font-bold mt-9" data-testid="login-heading">
        Login
      </h1>
      <form onSubmit={handleSubmit}>
        <div class="form-outline mb-4">
        <label class="form-label">
          User name
          </label>
          <input
            value={username}
            onChange={handleUsername}
            type="text"
            name="username"
            id="username"
            data-testid="username"
            class="form-control"
          />
          
        </div>

        <div class="form-outline mb-4">
        <label class="form-label">
            Password
          </label>
          <input
            type="password"
            value={password}
            onChange={handlePassword}
            name="password"
            id="password"
            data-testid="password"
            class="form-control"
          />
          
        </div>

        <button  class="btn btn-primary btn-block mb-4" type="submit" data-testid="submit">
          Sign in
        </button>
      </form>   
    </div>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  loggedIn: authedUser ? true : false,
});

export default connect(mapStateToProps)(Login);
