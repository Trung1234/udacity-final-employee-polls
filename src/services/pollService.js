import { _getUsers ,_getQuestions,_saveQuestion,_saveQuestionAnswer} from "../data/_DATA";
import { receiveQuestions } from "../store/actions/questions";
import { receiveUsers } from "../store/actions/users";


export function getInitialData () {
    return Promise.all([
        _getUsers(),
        _getQuestions(),
    ]).then(([users, questions]) => ({
        users,
        questions,
    }))
}

export function saveQuestion(optionOneText, optionTwoText, author) {
    return _saveQuestion({optionOneText, optionTwoText, author});
}

export function saveQuestionAnswer(authedUserId, qid, answer) {
    return _saveQuestionAnswer({
        authedUser: authedUserId,
        qid,
        answer
    });
}

export function handleInitialData() {
    return (dispatch) => {
        return getInitialData().then(({users, questions}) => {
            dispatch(receiveUsers(users));
            dispatch(receiveQuestions(questions));
        });
    };
}
