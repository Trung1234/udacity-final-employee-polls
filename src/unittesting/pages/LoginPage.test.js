
import LoginPage from "../../pages/LoginPage";
import { store } from "../../store";
import {fireEvent,  render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter} from "react-router-dom";
import React from "react";
import { handleInitialData } from "../../services/pollService";

describe("LoginPage", () => {
    it("should render the loginpage component", () => {
        const component = render (
            <Provider store={store}>
                <BrowserRouter>
                    <LoginPage/>
                </BrowserRouter>
            </Provider>
        );
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });

    it('should clear user and password after clicking submit button', async () => {
        await store.dispatch(handleInitialData());

        const wrapper = render(
            <Provider store={store}>
                <BrowserRouter>
                    <LoginPage/>
                </BrowserRouter>
            </Provider>
        );

        const loginHeadingElement = wrapper.getByTestId("employee-heading");
        const usernameInputElement = wrapper.getByTestId("username");
        const passwordInputElement = wrapper.getByTestId("password");
        const submitButtonElement = wrapper.getByTestId("submit");
        expect(loginHeadingElement).toBeInTheDocument();
        expect(usernameInputElement).toBeInTheDocument();
        expect(passwordInputElement).toBeInTheDocument();
        expect(submitButtonElement).toBeInTheDocument();

        fireEvent.change(usernameInputElement, {target: {value: 'test'}});
        fireEvent.change(passwordInputElement, {target: {value: 'wrongpassword'}});
        expect(usernameInputElement.value).toBe("test");
        expect(passwordInputElement.value).toBe("wrongpassword");
        fireEvent.click(submitButtonElement); 
        // confirm User stays on page
        expect(loginHeadingElement).toBeInTheDocument();
        expect(usernameInputElement.value).toBe("");
        expect(passwordInputElement.value).toBe("");
    });
});
