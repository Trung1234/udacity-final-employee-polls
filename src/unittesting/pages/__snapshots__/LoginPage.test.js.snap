// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should render the loginpage component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="container-fluid"
        >
          <div
            class="d-flex align-items-center justify-content-center h-100"
          >
            <img
              alt="employee"
              class="photo"
              data-testid="employee-heading"
              id="employee-heading"
              src="employee_surveys.jpg"
            />
          </div>
        </div>
        <form>
          <div
            class="form-outline mb-4"
          >
            <label
              class="form-label"
            >
              User name
            </label>
            <input
              class="form-control"
              data-testid="username"
              id="username"
              name="username"
              type="text"
              value="tylermcginnis"
            />
          </div>
          <div
            class="form-outline mb-4"
          >
            <label
              class="form-label"
            >
              Password
            </label>
            <input
              class="form-control"
              data-testid="password"
              id="password"
              name="password"
              type="password"
              value="abc321"
            />
          </div>
          <button
            class="btn btn-primary btn-block mb-4"
            data-testid="submit"
            type="submit"
          >
            Sign in
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="container-fluid"
      >
        <div
          class="d-flex align-items-center justify-content-center h-100"
        >
          <img
            alt="employee"
            class="photo"
            data-testid="employee-heading"
            id="employee-heading"
            src="employee_surveys.jpg"
          />
        </div>
      </div>
      <form>
        <div
          class="form-outline mb-4"
        >
          <label
            class="form-label"
          >
            User name
          </label>
          <input
            class="form-control"
            data-testid="username"
            id="username"
            name="username"
            type="text"
            value="tylermcginnis"
          />
        </div>
        <div
          class="form-outline mb-4"
        >
          <label
            class="form-label"
          >
            Password
          </label>
          <input
            class="form-control"
            data-testid="password"
            id="password"
            name="password"
            type="password"
            value="abc321"
          />
        </div>
        <button
          class="btn btn-primary btn-block mb-4"
          data-testid="submit"
          type="submit"
        >
          Sign in
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
